### PR TITLE
Add message to task queue when file is uploaded

### PR DESCRIPTION
**Pull request by zekehuntergreen**
## What does this change?
- updates the `transcribe-file` api endpoint (renamed from `send-message`) so that it 
  - takes an  `s3Key` parameter 
  - creates a presigned url to download the file with the given key in the uploads bucket
  - adds a message to the task queue (with non-test data) including this url so that the worker can use it to get the file
- change to client so that when the user submits the form, after the file is uploaded to s3, `transcribe-file` is called.
- change to worker so that it uses this presigned url rather than having access to the entire bucket. This change is important because it will allow third parties to integrate with the transcription service without our having to change any of the worker permissions. 

## How to test
- can be tested locally by starting api, client, worker
- testing in CODE

## Accessibility

-   [ ] [Tested with screen reader](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#screen-reader)
-   [ ] [Navigable with keyboard](https://github.com/guardian/accessibility/blob/main/people-and-technology/02-physical.md#keyboard)
-   [ ] [Colour contrast passed](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#contrast)
-   [ ] [The change doesn't use only colour to convey meaning](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#use-of-colour)

### CODE REVIEW

File: packages/api/src/index.ts [hunk 1]
```diff
@@ -61,21 +63,33 @@ const getApp = async () => {
...
+			const userEmail = req.user?.email;
+			console.log(req.body);
+			const body = sendMessageRequestBody.safeParse(req.body);
+			if (!body.success || !userEmail) {
+				console.error(body, userEmail);
+				res.status(422).send('missing request params');
+				return;
+			}
+
+			const s3Key = body.data.s3Key;
```
Comment (zekehuntergreen) [inline packages/api/src/index.ts:78]: need to check that current user was the one to upload the file with this key

File: packages/backend-common/src/s3.ts [hunk 2]
```diff
@@ -57,28 +70,52 @@ export const getFile = async (
...
+	await downloadS3Data(body, destinationPath, key);
+	return destinationPath;
+};
+
+const downloadS3Data = async (
+	data: Readable,
+	destinationPath: string,
+	key: string,
+) => {
+	const body = ReadableBody.parse(data);
```
Comment (marjisound) [inline packages/backend-common/src/s3.ts:100]: It seems that we have already parsed the sream into `ReadableBody` on line 90. Probably this is redundant here?

**Review by marjisound — COMMENTED** [review]
Looks great, I just had 2 minor questions

File: packages/backend-common/src/s3.ts [hunk 3]
```diff
@@ -32,15 +31,29 @@ export const getSignedUrl = (
...
+			Key: id,
 			Metadata: {
 				'user-email': userEmail,
-				'file-name': fileName,
 			},
 		}),
 		{ expiresIn }, // override default expiration time of 15 minutes
 	);
 
+export const getSignedDownloadUrl = async (
```
Comment (philmcmahon) [inline packages/backend-common/src/s3.ts:42]: slightly out of the scope of this pr but I think worth doing - could generate the s3 client at the top level in the various index.ts files and then pass it through to this function and getSignedUploadurl.

You could potentially reduce code duplication even more by having just one function that either takes a GetObjectCommand or a PutObjectCommand

File: packages/api/src/index.ts [hunk 4]
```diff
@@ -147,25 +161,21 @@ const getApp = async () => {
...
+			const s3Key = uuid4();
+			const presignedS3Url = await getSignedUploadUrl(
 				config.aws.region,
 				config.app.sourceMediaBucket,
 				req.user?.email ?? 'not found',
-				queryParams.data.fileName,
 				60,
 				true,
+				s3Key,
 			);
```
[the remark this answers is not in the document]
Comment (zekehuntergreen) [reply packages/api/src/index.ts:175]: from conversation in standup last week - we've decided we're ok with orphaned files in the media bucket when the upload to the bucket succeeds but the POST to transcribe-file fails. The user will just have to try uploading again

**Review by marjisound — APPROVED** [review]
Looks great 🎉

**Pull request merged**